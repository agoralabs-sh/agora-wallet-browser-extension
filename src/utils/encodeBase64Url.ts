/**
 * Convenience function that encodes a string or arbitrary bit of data to a base64 url string
 * @param {string | Uint8Array} input - the string or data to encode.
 * @returns {string} a base64 Url encoded string.
 * @throws {InvalidCharacterError} if the input contains characters that did not fit in a single byte.
 */
export default function encodeBase64Url(input: string | Uint8Array): string {
  const parseToBase64Url: (base64String: string) => string = (
    base64String: string
  ) => base64String.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_');
  let decoder: TextDecoder;
  let decodedString: string;

  if (typeof input === 'string') {
    return parseToBase64Url(window.btoa(input));
  }

  decoder = new TextDecoder();
  decodedString = decoder.decode(input);

  return parseToBase64Url(window.btoa(decodedString));
}
