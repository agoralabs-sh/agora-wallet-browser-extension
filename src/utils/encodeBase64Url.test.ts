import { randomBytes } from 'tweetnacl';

// Utils
import encodeBase64Url from './encodeBase64Url';

describe('encodeBase64Url()', () => {
  it('should successfully encode a string', () => {
    // Arrange
    const input: string = 'Hello human!';
    const expected: string = Buffer.from(input).toString('base64url');
    // Act
    const result: string = encodeBase64Url(input);

    // Assert
    expect(result).toBe(expected);
  });

  it('should throw an error if the input is not a unicode character', () => {
    // Arrange
    // Act
    try {
      encodeBase64Url(randomBytes(32));
    } catch (error) {
      // Assert
      expect(error).toMatchSnapshot();

      return;
    }

    throw new Error('should throw an InvalidCharacterError');
  });

  it('should successfully encode data', () => {
    // Arrange
    const encoder: TextEncoder = new TextEncoder();
    const input: string = 'Hello human!';
    const expected: string = Buffer.from(input).toString('base64url');
    // Act
    const result: string = encodeBase64Url(encoder.encode(input));

    // Assert
    expect(result).toBe(expected);
  });
});
