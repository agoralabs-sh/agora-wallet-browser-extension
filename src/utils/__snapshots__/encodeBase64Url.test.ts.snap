// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encodeBase64Url() should throw an error if the input is not a unicode character 1`] = `[InvalidCharacterError: The string to be encoded contains invalid characters.]`;
